distributed/worker.py

```diff
@@ -3149,38 +3151,35 @@ def _put_key_in_memory(self, ts: TaskState, value, *, stimulus_id: str) -> Recs:
         return recommendations
 
     def _select_keys_for_gather(
-        self, worker: str, dep: str, all_keys_to_gather: Container[str]
+        self, worker: str, ts: TaskState
     ) -> tuple[set[str], int]:
         """``_ensure_communicating`` decided to fetch a single task from a worker,
         following priority. In order to minimise overhead, request fetching other tasks
         from the same worker within the message, following priority for the single
         worker but ignoring higher priority tasks from other workers, up to
         ``target_message_size``.
         """
-        deps = {dep}
-
-        total_bytes = self.tasks[dep].get_nbytes()
+        keys = {ts.key}
+        total_bytes = ts.get_nbytes()
         tasks = self.data_needed_per_worker[worker]
 
         while tasks:
             ts = tasks.peek()
-            if (
-                ts.state != "fetch"
-                # Do not acquire the same key twice if multiple workers holds replicas
-                or ts.key in all_keys_to_gather
-                # A replica is still available (otherwise status would not be 'fetch'
-                # anymore), but not on this worker. See _update_who_has().
-                or worker not in ts.who_has
-            ):
-                tasks.pop()
-                continue
+            if self.validate:
+                assert ts.state == "fetch"
+                assert worker in ts.who_has
             if total_bytes + ts.get_nbytes() > self.target_message_size:
                 break
             tasks.pop()
-            deps.add(ts.key)
+            self.data_needed.remove(ts)
+            for other_worker in ts.who_has:
+                if other_worker != worker:
+                    self.data_needed_per_worker[other_worker].remove(ts)
+
+            keys.add(ts.key)
             total_bytes += ts.get_nbytes()
 
-        return deps, total_bytes
+        return keys, total_bytes
```

Comment:
Why return keys here instead of `TaskState` instances?

Reply:
Convenience due to how the output is used. But I agree it's slightly cleaner the other way around; changing it.